### PR TITLE
[model, rollout, reward, algo, worker, trainer] feat: Qwen3-TTS talker GSPO + online DPO

### PR DESCRIPTION
**Pull request by williammtan**
### What does this PR do?

Adds native Qwen3-TTS talker RL to verl-omni as two recipes that share one talker actor and one
vLLM-Omni AR rollout:

- **GSPO** (group RL): a multi-metric audio reward (intelligibility, speaker similarity, emotion, stability) through verl's stock GSPO policy loss and GDPO advantage estimator.
- **Online DPO**: two candidates per prompt, judged pairwise by a remote audio-capable LLM (Gemini by default, any endpoint), trained with a pairwise DPO loss on the talker's codec-0 sequence log-prob against a frozen reference.

Addresses RFC #90. The talker is not a plain causal LM (2-channel input, speaker slot, 16 codebooks), so the actor computes real teacher-forced codec-0 log-probs and the rollout stays on-policy while the codec2wav decoding is reward-side.

### Checklist Before Starting

- [x] Searched for similar open PRs: [`is:pr is:open tts`](https://github.com/verl-project/verl-omni/pulls?q=is%3Apr+is%3Aopen+tts), plus `qwen3-tts`, `speechjudge`, `audio judge dpo`. No open PR contains Qwen3-TTS or TTS RL (the single `tts` hit, #233, is a CI refactor). RFC #90 requests this recipe and has no implementation PR.
- [x] Title formatted as `[{modules}] {type}: {description}`: `[model, rollout, reward, algo, worker, trainer] feat: Qwen3-TTS talker GSPO + online DPO`.
- [x] No breaking API change: everything is additive (a new `[tts]` extra, new `reward.*` / `actor.policy_loss` config keys, a new trainer entry). Existing T2I/LLM/diffusion behavior is untouched.

#### CPU tests

```bash
pytest -q --asyncio-mode=auto tests/     # python_files = *_on_cpu.py
```

```text
299 passed   # 45 in the 7 TTS-specific files
```

| Area | File |
| --- | --- |
| talker forward (layout; verl-gather == talker-gather) | `tests/models/test_qwen3_tts_forward_on_cpu.py` |
| reward functions (asr/sim/emo, CER, stability, composition) | `tests/utils/reward_score/test_tts_rewards_on_cpu.py` |
| audio reward managers (decode, multi weighted-sum) | `tests/reward_loop/test_audio_reward_manager_on_cpu.py` |
| audio-judge rendezvous + `num_workers=1` guard | `tests/reward_loop/test_audio_judge_reward_manager_on_cpu.py` |
| DPO loss + pairing + ties + NLL anchor | `tests/workers/test_tts_dpo_loss_on_cpu.py` |
| `DPOPolicyLossConfig` via `_target_` | `tests/workers/config/test_dpo_policy_loss_config_on_cpu.py` |
| TTS-mode detection from stage config | `tests/workers/rollout/rollout_vllm/test_tts_mode_detection_on_cpu.py` |

#### GPU end-to-end (1 node x 8 H200)

| Recipe | Run | Result |
| --- | --- | --- |
| GSPO | 25-step regression vs a reference run | EXIT=0; rollout/actor Pearson 0.9993; reward dims match reference (asr 0.99, sim 0.884, emo 0.14, stab 0.0) |
| DPO | 2-step on pristine base verl + a mock `/rank` judge | EXIT=0; `actor/dpo_loss = 0.6931 = ln 2` at init (policy == reference, so the margin is 0), which confirms the loss bind, `ref_log_prob`, and pairing are all live; Pearson 0.9993 |
| GSPO via the shared `main_tts` entry | 2-step | EXIT=0; metric is `actor/pg_loss` (not `actor/dpo_loss`), so the DPO bind is a no-op for GSPO; numbers identical to the base-entry run |

GSPO was also validated originally with a full training pass over a ~25k-line voice dataset (all reward dimensions rise). Caveat: an LLM judge discriminates weakly on same-text / same-voice pairs, so online DPO from it can regress on quality even while training metrics look healthy; evaluate checkpoints on held-out data (documented in the DPO README).

#### Static checks

`ruff check` / `ruff format`, `tests/special_sanity/check_license.py`, and
`tests/special_sanity/validate_structure.py` pass.

### API and Usage Example

GSPO:

```bash
python data_process/tts_content_synth.py --input conversations.jsonl --ref_audio clone.wav --output_dir ~/data/tts
python data_process/precompute_spk_embed.py --ref clone.wav --out ~/data/tts/spk_embed.json

MODEL_PATH=<qwen3-tts ckpt> TRAIN_FILE=~/data/tts/train.parquet VAL_FILE=~/data/tts/test.parquet \
  SPK_EMBED=~/data/tts/spk_embed.json \
  bash examples/qwen3_tts_gspo_trainer/run_qwen3_tts_gspo.sh
```

Online DPO (start a judge, then train):

```bash
GOOGLE_API_KEY=... python examples/qwen3_tts_dpo/judge_server.py --provider gemini --port 8901

MODEL_PATH=<qwen3-tts SFT ckpt> TRAIN_FILE=... VAL_FILE=... SPK_EMBED=... JUDGE_URL=http://localhost:8901 \
  bash examples/qwen3_tts_dpo/run_qwen3_tts_dpo.sh
```

Both launch `python3 -m verl_omni.trainer.main_tts`. Install adds the `[tts]` extra plus `pip install --no-deps qwen-tts==0.1.1 sox`; three idempotent dependency-wall patches ship in `examples/qwen3_tts_gspo_trainer/patches/` (documented in the README).

### Design & Code Changes

- **Talker actor** — `verl_omni/models/transformers/qwen3_tts.py` (external_lib patch: registers the
  qwen-tts classes, installs the teacher-forced codec-0 forward, talker-only freeze) +
  `qwen3_tts_forward.py` (side-effect-free tensor math, CPU-tested).
- **Rollout** — `verl_omni/workers/rollout/vllm_rollout/vllm_omni_async_server.py`: voice-clone AR
  generation; TTS mode detected once from the stage's `codec` output type; codec frames surfaced as
  `tts_audio_codes` for the actor's recompute.
- **Reward** — `reward_loop/reward_manager/{audio,multi,audio_judge}.py` and
  `utils/reward_score/tts/`: `AudioRewardManager` mirrors `VisualRewardManager`;
  `MultiAudioRewardManager` composes per-model reward functions; `AudioJudgeRewardManager` adds the
  DPO judge rendezvous. `examples/qwen3_tts_dpo/judge_server.py` is a small generic Gemini /
  OpenAI-compatible `/rank` server.
- **DPO** — `workers/utils/losses.py` (`tts_dpo_loss`, pairs chosen/rejected internally),
  `workers/config/tts/actor.py` (`DPOPolicyLossConfig` carries β/λ via `_target_`), and
  `trainer/main_tts.py` (the shared Qwen3-TTS entry: base verl's V0 runner verbatim plus one
  `set_loss_fn` call, a no-op unless `loss_mode` is dpo).

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md).
- [x] Ran the pre-commit hooks (ruff check/format, license header, structure validation) and the CPU suite; all pass.
- [x] Added docs: `examples/qwen3_tts_gspo_trainer/README.md` and `examples/qwen3_tts_dpo/README.md`.
- [x] Added CPU tests (`*_on_cpu.py`, auto-collected by the [CPU CI workflow](https://github.com/verl-project/verl-omni/tree/main/.github/workflows)). GPU end-to-end needs an 8xH200 node and a qwen-tts checkpoint, so it runs manually and is not wired into CI.

### AI Assistance

AI assistance was used to restructure the implementation from an internal branch, add the tests, and draft this description. I reviewed the final diff and the test results and can explain and maintain the change end to end.

### CODE REVIEW

**Comment by SamitHuang** [comment]
@wtomin  PTAL at the DPO part

File: examples/qwen3_tts_gspo_trainer/patches/patch_qwen_tts_tf5.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    "    import torch\n"
+    "    base = getattr(config, 'rope_theta', 10000.0) or 10000.0\n"
+    "    dim = getattr(config, 'head_dim', None) or (config.hidden_size // config.num_attention_heads)\n"
+    "    inv_freq = 1.0 / (base ** (torch.arange(0, dim, 2, dtype=torch.int64)"
+    ".to(device=device, dtype=torch.float) / dim))\n"
+    "    return inv_freq, 1.0\n\n\n"
+)
+
+
+def patch_file(path):
```
Comment (Luosuu) [inline examples/qwen3_tts_gspo_trainer/patches/patch_qwen_tts_tf5.py:56]: this is too hacking. consider better approaches. maybe we can support such utils in veomni modeling

File: examples/qwen3_tts_gspo_trainer/README.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
+# Qwen3-TTS Talker GSPO Trainer
+
+RL post-training of `Qwen/Qwen3-TTS-12Hz-1.7B-Base`: the talker learns to speak text lines
+better, scored by audio reward functions on the decoded waveform. FSDP actor (talker codec-0
+path) + vLLM-Omni AR rollout, GSPO token-level policy loss + GDPO per-dimension advantage
+estimator (both from base verl).
+
+Validated end-to-end with a full training pass over a ~25k-line voice dataset on 1 node x 8 H200
+(reward rises on all dimensions; rollout and actor codec log-probs match to Pearson >= 0.999).
```
Comment (SamitHuang) [inline examples/qwen3_tts_gspo_trainer/README.md:9]: Pls attach the validation results (reward curves etc) to verify the convergence in the PR description and update [performance reference](https://verl-omni.readthedocs.io/en/latest/algo/performance.html). thx

File: examples/qwen3_tts_gspo_trainer/README.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+On top of the standard verl-omni GPU install, the recipe needs the TTS reward extras and the
+qwen-tts package (reward-side code2wav + the trainable model classes):
+
+```bash
+pip install -e ".[tts]"
+pip install --no-deps qwen-tts==0.1.1 sox
+```
+
+qwen-tts 0.1.1 pins transformers==4.57.3, which vllm 0.24 no longer supports; --no-deps plus a
```
Comment (SamitHuang) [inline examples/qwen3_tts_gspo_trainer/README.md:82]: Is it possible to adapt qwen3_tts implementation to `transformers` 5.x? It's better to drop redundant patches for easier maintenance.

Comment (SamitHuang) [reply examples/qwen3_tts_gspo_trainer/README.md:82]: Qwen3Omni also faced the transformers version issure previsouly. It's better to create another PR for transformers 5.x adaptation, referring to https://github.com/verl-project/verl-omni/pull/208

**Review by SamitHuang — COMMENTED** [review]
Thanks for the great effort. Overall it's good and the training results look solid. Just need to enhance the code and documentation a bit to ensoure open-source quality before merge.

1. updated the docs for support models https://verl-omni.readthedocs.io/en/latest/start/models.html
2. it's better to update the design docs (How to add a text-to-audio model), or RFC(https://github.com/verl-project/verl-omni/issues/90) for the current implementation such as the added MultiAudioRewardManager and AudioRewardManager

**Review by SamitHuang — COMMENTED** [review]
btw, since this PR is quite large, how about splitting it into at least 2 PRs, one for qwen-tts DPO and another for GSPO?

File: pyproject.toml [hunk 4]
```diff
@@ -73,6 +73,19 @@ dev = [
...
+# hard-pins transformers==4.57.3, so install it separately with --no-deps; see
+# examples/qwen3_tts_gspo_trainer/README.md.
+tts = [
+    "librosa",
+    "soundfile",
+    "faster-whisper",
+    "modelscope",
+    "addict",
+    "funasr",
+    "simplejson",  # undeclared modelscope import; without it the ERes2Net load fails silently
```
[the remark this answers is not in the document]
Comment (SamitHuang) [reply pyproject.toml:88]: agree

File: examples/qwen3_tts_gspo_trainer/patches/patch_verl_unpad_fallback.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    "        try:\n"
+    "            from flash_attn.bert_padding import index_first_axis, pad_input, rearrange, unpad_input\n"
+    "        except ImportError:\n"
+    "            from einops import rearrange\n"
+    "            from transformers.modeling_flash_attention_utils import (\n"
+    "                _index_first_axis as index_first_axis,\n"
+    "                _pad_input as pad_input,\n"
+    "                _unpad_input as unpad_input,\n"
+    "            )"
+)
```
Comment (SamitHuang) [inline examples/qwen3_tts_gspo_trainer/patches/patch_verl_unpad_fallback.py:44]: Too hacky. we can implement it in `utils/attention_utils.py` in verl-omni to handle the difffernt attention backends(FA3, FA2, SPDA) for the actor. PTAL @zhtmike

**Comment by zhtmike** [comment]
Can you show the reward curve?

**Comment by SamitHuang** [comment]
@williammtan  PTAL

**Comment by wtomin** [comment]
@williammtan Hello. I think this PR is a bit too large. Could you split it to two PRs, one for GSPO and one for online DPO?

**Comment by wtomin** [comment]
@williammtan  Hello, William. Is this PR still active?
